File: benchmarks/31-nintegrate-ladder/nintegrate_ladder.m
(* Experiment 31 -- NIntegrate, a difficulty ladder.
   ==========================================================================
   WHY A LADDER.  Experiment 12 measures four integrand classes at one
   difficulty each, which answers "is the quadrature there" but not "where does
   it break".  An adaptive rule does not degrade smoothly: it is flat until the
   integrand out-runs the rule's assumption, then falls off a cliff.  Every
   family below is therefore run at INCREASING difficulty along its own axis,
   so a cliff shows up as a jump between two adjacent rows of the same family
   rather than as one number with nothing to compare it to.

   THE AXES
     oscillation   Sin[k x], k = 40 -> 200 -> 1000.  Cost should grow like the
                   number of half-periods unless a dedicated oscillatory rule
                   is reached; a rule that merely bisects pays much more.
     peak width    Exp[-w (x-1/2)^2], w = 1e3 -> 1e5.  The peak occupies
                   1/Sqrt[w] of the interval, so a bisecting rule must find a
                   feature 10x narrower in the second row than the first.
     endpoint      x^-1/2 -> x^-9/10.  Both integrable, the second far closer
                   to divergence; this is what double-exponential is for.
     dimension     1-D -> 2-D -> 3-D.  Cubature cost is exponential in d, so
                   these rows say what the base is.
     precision     machine -> 30 digits -> 50 digits.  This is where a fixed
                   double-precision rule has to hand over to MPFR.

   CHECKS ARE ACCURACY, NOT AGREEMENT.  Every integrand here has a known
   closed-form value, and each check rounds the computed result against it. A
   system that is fast because it stopped early fails the check rather than
   winning the row -- the distinction the ranked report cannot otherwise make.
   ========================================================================== *)

Get["../harness.m"];
Get["../data.m"];

require[{"NIntegrate", "WorkingPrecision", "AccuracyGoal", "PrecisionGoal"}];

(* ---- oscillation: Integrate[Sin[k x], {x, 0, Pi}] = (1 - Cos[k Pi])/k ----
   k even => exactly 0, so the check is the SCALED value: rounding a true zero
   to four places would pass for any answer under 5e-5 and measure nothing. *)
bench["NI oscillatory k=40", NIntegrate[Sin[40 x], {x, 0, Pi}];];
check["NI oscillatory k=40", Round[10^6 NIntegrate[Sin[40 x], {x, 0, Pi}]]];

bench["NI oscillatory k=200", NIntegrate[Sin[200 x], {x, 0, Pi}];];
check["NI oscillatory k=200", Round[10^6 NIntegrate[Sin[200 x], {x, 0, Pi}]]];

bench["NI oscillatory k=1000", NIntegrate[Sin[1000 x], {x, 0, Pi}];];
check["NI oscillatory k=1000", Round[10^6 NIntegrate[Sin[1000 x], {x, 0, Pi}]]];

(* Odd k, so the answer is 2/k rather than 0 -- a nonzero oscillatory value
   the check can actually pin down. *)
bench["NI oscillatory k=1001 nonzero", NIntegrate[Sin[1001 x], {x, 0, Pi}];];
check["NI oscillatory k=1001 nonzero",
  Round[10^6 NIntegrate[Sin[1001 x], {x, 0, Pi}]]];

(* ---- peak width: Integrate[Exp[-w (x-1/2)^2], {x,0,1}] -> Sqrt[Pi/w] ----
   The ladder stops at w = 1e5. At w = 1e6 Mathematica returns 0.000886, i.e.
   exactly HALF the correct 0.00177245, having resolved one side of the peak
   and not the other; Mathilda and scipy both get it right. That is a real
   finding, but as a row it would only ever be a check disagreement, and the
   harness withholds timings for those -- so it is recorded here rather than
   left as a permanently red row measuring nothing. *)
bench["NI narrow peak w=1e3", NIntegrate[Exp[-1000 (x - 1/2)^2], {x, 0, 1}];];
check["NI narrow peak w=1e3",
  Round[10^8 NIntegrate[Exp[-1000 (x - 1/2)^2], {x, 0, 1}]]];

bench["NI narrow peak w=1e5",
  NIntegrate[Exp[-100000 (x - 1/2)^2], {x, 0, 1}];];
check["NI narrow peak w=1e5",
  Round[10^8 NIntegrate[Exp[-100000 (x - 1/2)^2], {x, 0, 1}]]];

(* ---- endpoint singularity: 1/Sqrt[x] -> 2, x^(-9/10) -> 10 ---- *)
bench["NI endpoint x^-1/2", NIntegrate[1/Sqrt[x], {x, 0, 1}];];
check["NI endpoint x^-1/2", Round[10^6 NIntegrate[1/Sqrt[x], {x, 0, 1}]]];

bench["NI endpoint x^-9/10", NIntegrate[x^(-9/10), {x, 0, 1}];];
check["NI endpoint x^-9/10", Round[10^4 NIntegrate[x^(-9/10), {x, 0, 1}]]];

(* ---- interior kink: Integrate[Abs[x - 1/3], {x,0,1}] = 5/18 ----
   Not a singularity, but the derivative jumps, so an error estimate built on a
   smoothness assumption is wrong exactly at the panel containing 1/3. *)
bench["NI interior kink", NIntegrate[Abs[x - 1/3], {x, 0, 1}];];
check["NI interior kink", Round[10^6 NIntegrate[Abs[x - 1/3], {x, 0, 1}]]];

(* ---- semi-infinite range: Integrate[Exp[-x^2], {x,0,Infinity}] = Sqrt[Pi]/2 *)
bench["NI semi-infinite Gaussian", NIntegrate[Exp[-x^2], {x, 0, Infinity}];];
check["NI semi-infinite Gaussian",
  Round[10^6 NIntegrate[Exp[-x^2], {x, 0, Infinity}]]];

(* ---- dimension: 2-D and 3-D cubature, both with closed forms ----
   2-D: Integrate[x y, {x,0,1},{y,0,1}] = 1/4;  3-D: x y z -> 1/8. *)
bench["NI 2-D product", NIntegrate[x y, {x, 0, 1}, {y, 0, 1}];];
check["NI 2-D product", Round[10^6 NIntegrate[x y, {x, 0, 1}, {y, 0, 1}]]];

bench["NI 3-D product", NIntegrate[x y z, {x, 0, 1}, {y, 0, 1}, {z, 0, 1}];];
check["NI 3-D product",
  Round[10^6 NIntegrate[x y z, {x, 0, 1}, {y, 0, 1}, {z, 0, 1}]]];

(* 2-D with a ridge rather than a product: no separability to exploit. *)
bench["NI 2-D ridge", NIntegrate[Exp[-100 (x - y)^2], {x, 0, 1}, {y, 0, 1}];];
check["NI 2-D ridge",
  Round[10^6 NIntegrate[Exp[-100 (x - y)^2], {x, 0, 1}, {y, 0, 1}]]];

(* ---- precision: machine -> 30 -> 50 digits ----
   Integrate[Sin[x], {x,0,Pi}] = 2 exactly, so the check is the same value at
   every precision and any digit lost is the row's own fault. *)
bench["NI 30-digit Sin", NIntegrate[Sin[x], {x, 0, Pi}, WorkingPrecision -> 30];];
check["NI 30-digit Sin",
  Round[10^6 NIntegrate[Sin[x], {x, 0, Pi}, WorkingPrecision -> 30]]];

bench["NI 50-digit Gaussian",
  NIntegrate[Exp[-x^2], {x, 0, 3}, WorkingPrecision -> 50];];
check["NI 50-digit Gaussian",
  Round[10^6 NIntegrate[Exp[-x^2], {x, 0, 3}, WorkingPrecision -> 50]]];

File: benchmarks/31-nintegrate-ladder/nintegrate_ladder.py
#!/usr/bin/env python3
"""Experiment 31 -- NIntegrate difficulty ladder (scipy.integrate column).

Same rows as ``nintegrate_ladder.m``, same order, same labels.

WHAT THE BASELINE IS.  scipy's ``quad`` is QUADPACK, i.e. adaptive
Gauss-Kronrod with the same lineage as the CAS columns' default rule, so these
rows compare like with like -- unlike, say, the graph experiment, where the
baseline is pure Python and a win means less.

WHERE THE BASELINE IS TOLD THE ANSWER.  QUADPACK needs help on two families,
and giving it that help is the honest comparison, because a user of scipy would
give it too:

  * oscillatory rows use ``limit`` raised well above the default 50 subdivisions
    -- without it QUADPACK returns a convergence warning rather than a number
    at k = 1000;
  * the kinked integrand is handed ``points=[1/3]`` so the panel boundary lands
    on the kink, which is what the argument exists for.

The peak and endpoint rows are deliberately NOT helped: locating a narrow peak
and handling an integrable endpoint singularity are what the adaptive rule and
the double-exponential transform respectively claim to do unaided, so helping
them would measure the wrong thing.

High-precision rows use mpmath, since float64 cannot represent the answer at
all -- that is the point of those two rows.
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import math

import mpmath
from scipy.integrate import quad, dblquad, tplquad

from harness import bench, check, require

require(["scipy.integrate:quad", "scipy.integrate:dblquad",
         "scipy.integrate:tplquad", "mpmath:quad"])


def r(x, p):
    """Round like the .m column's Round[10^p x] -- half away from zero."""
    v = x * (10 ** p)
    return int(math.floor(v + 0.5)) if v >= 0 else -int(math.floor(-v + 0.5))


# ---- oscillation ---------------------------------------------------------
# limit=2000: at k=1000 the default 50 subdivisions cannot resolve 500 periods
# and QUADPACK reports non-convergence instead of an answer.
def osc(k):
    return quad(lambda x: math.sin(k * x), 0.0, math.pi, limit=2000)[0]


bench("NI oscillatory k=40", lambda: osc(40))
check("NI oscillatory k=40", r(osc(40), 6))

bench("NI oscillatory k=200", lambda: osc(200))
check("NI oscillatory k=200", r(osc(200), 6))

bench("NI oscillatory k=1000", lambda: osc(1000))
check("NI oscillatory k=1000", r(osc(1000), 6))

bench("NI oscillatory k=1001 nonzero", lambda: osc(1001))
check("NI oscillatory k=1001 nonzero", r(osc(1001), 6))


# ---- peak width ----------------------------------------------------------
def peak(w):
    return quad(lambda x: math.exp(-w * (x - 0.5) ** 2), 0.0, 1.0)[0]


bench("NI narrow peak w=1e3", lambda: peak(1e3))
check("NI narrow peak w=1e3", r(peak(1e3), 8))

bench("NI narrow peak w=1e5", lambda: peak(1e5))
check("NI narrow peak w=1e5", r(peak(1e5), 8))


# ---- endpoint singularity ------------------------------------------------
bench("NI endpoint x^-1/2", lambda: quad(lambda x: x ** -0.5, 0.0, 1.0)[0])
check("NI endpoint x^-1/2", r(quad(lambda x: x ** -0.5, 0.0, 1.0)[0], 6))

bench("NI endpoint x^-9/10", lambda: quad(lambda x: x ** -0.9, 0.0, 1.0)[0])
check("NI endpoint x^-9/10", r(quad(lambda x: x ** -0.9, 0.0, 1.0)[0], 4))


# ---- interior kink -------------------------------------------------------
def kink():
    return quad(lambda x: abs(x - 1.0 / 3.0), 0.0, 1.0, points=[1.0 / 3.0])[0]


bench("NI interior kink", kink)
check("NI interior kink", r(kink(), 6))


# ---- semi-infinite range -------------------------------------------------
def semiinf():
    return quad(lambda x: math.exp(-x * x), 0.0, math.inf)[0]


bench("NI semi-infinite Gaussian", semiinf)
check("NI semi-infinite Gaussian", r(semiinf(), 6))


# ---- dimension -----------------------------------------------------------
def d2():
    return dblquad(lambda y, x: x * y, 0.0, 1.0, 0.0, 1.0)[0]


def d3():
    return tplquad(lambda z, y, x: x * y * z, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0)[0]


def ridge():
    return dblquad(lambda y, x: math.exp(-100.0 * (x - y) ** 2),
                   0.0, 1.0, 0.0, 1.0)[0]


bench("NI 2-D product", d2)
check("NI 2-D product", r(d2(), 6))

bench("NI 3-D product", d3)
check("NI 3-D product", r(d3(), 6))

bench("NI 2-D ridge", ridge)
check("NI 2-D ridge", r(ridge(), 6))


# ---- precision -----------------------------------------------------------
# float64 has ~15.9 digits, so these two rows are mpmath's by necessity, not by
# preference: the .m columns are being asked for 30 and 50 correct digits.
def mp_sin30():
    with mpmath.workdps(30):
        return mpmath.quad(mpmath.sin, [0, mpmath.pi])


def mp_gauss50():
    with mpmath.workdps(50):
        return mpmath.quad(lambda t: mpmath.exp(-t * t), [0, 3])


bench("NI 30-digit Sin", mp_sin30)
check("NI 30-digit Sin", r(float(mp_sin30()), 6))

bench("NI 50-digit Gaussian", mp_gauss50)
check("NI 50-digit Gaussian", r(float(mp_gauss50()), 6))
